### PR TITLE
[SPARK-13244] [WIP] DataFrame to Dataset migration

### PR DESCRIPTION
**Pull request by liancheng**
This PR prototypes the `DataFrame`-to-`Dataset` migration.  It demonstrates main migration strategies but hasn't migrate all `DataFrame` methods.  Compilation/tests may break.

Goals:
1.  `Dataset` should have all operations of the original `DataFrame`
2.  `DataFrame` is just a special `Dataset[Row]`
3.  Source level compatibility whenever possible

Basic strategies used:
1.  Migrate all methods in `DataFrame` into `Dataset`
2.  Migrate all methods in `GroupedData` into `GroupedDataset[K, V]` and then remove `GroupedData`
3.  For all migrated untyped methods (typically those take `String`s and `Column`s as arguments), make them return `Row`, `Dataset[Row]`, or `GroupedDataset[Row, T]`.
4.  Make `DataFrame` extend from `Dataset[Row]`
   
   Overrides untyped `Dataset` operations in `DataFrame` and make them return `DataFrame` instead of `Dataset[Row]`

Problems encountered:
1.  `DataFrame` RDD operations conflict with `Dataset` functional operations
   
   E.g.: `df.map` returns an RDD while `df.map` returns a `Dataset`.
   
   Possible solutions:
   -   Remove `DataFrame` RDD operations since they can be easily mapped to methods of `df.rdd.
     
     Cons: breaks source level compatibility
   -   Rename those `Dataset` methods
     
     Cons: may result inconsistent naming
   
   Personally I prefer the former option.
2.  `Dataset` methods returning arrays can't be overriden in `DataFrame`
   
   E.g.: `Dataset.randomSplit` returns `Array[Dataset[T]]`.  We can't have an override version for `DataFrame` and make it return `Array[DataFrame]`.

### CODE REVIEW

**Comment by rxin** [comment]
What's the reason we need mapRows? Is it to keep binary compatibility for the map function?

**Comment by liancheng** [comment]
It's because of the problem 1 mentioned in the PR description. Renaming `DataFrame.map` to `DataFrame.mapRows` is to avoid conflict with `Dataset.map` since now `DataFrame` extends from `Dataset[Row]`.

Actually I tend to remove these methods and simply resort to `df.rdd.map` etc.. However, renaming them is more convenient with the help of IDE refactoring, otherwise I'll have to update the whole codebase manually. Would like to get everything work first. May remove them later.

**Comment by rxin** [comment]
OK that makes sense. We shouldn't have mapRows.

**Comment by liancheng** [comment]
One super annoying thing is that IDEA turns crazy and highlights tons of legal code as red after making `DataFrame` extend from `Dataset[Row]`...

**Comment by rxin** [comment]
OK I suggest breaking this into two changes.

The first one replaces all the df.map call with df.rdd.map (and flatMap, mapPartitions, etc).

File: mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala [hunk 1]
```diff
@@ -59,6 +59,7 @@ private[classification] trait LogisticRegressionParams extends ProbabilisticClas
    *       equivalent.
    *
    * Default is 0.5.
+ *
```
Comment (liancheng) [inline mllib/src/main/scala/org/apache/spark/ml/classification/LogisticRegression.scala]: IDEA Scala plugin messes up with ScalaDoc when doing refactoring and import organization. Will revert these unexpected changes.

**Comment by liancheng** [comment]
This PR is only used for prototyping. I'm closing it and will open follow-up PRs to complete it.

**Pull request closed without merging**